### PR TITLE
Amazon Q Transform: refactor gumby client and messages

### PR DESCRIPTION
**Pull request by tincheng**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

1. Add back logging for general errors caught when failing to upload artifact or start job, and fix logging level for multiple informational or error logs.
2. Refactor GumbyClient with a generic `callApi()` function that handles logging and metrics for all client API call. 
 a. The only exception being `UploadZip` which is not an actual API call, but since we are currently emitting API latency/error metrics for it, will keep the metrics for consistency.
 b. No longer tracking uploadId in latency metric for `CreateUploadUrl` API, since uploadId is an output and not an input for that API call. 
 c. No longer tracking byteSize in latency metric for `ExportResultArchive` API, since byteSize is already tracked in existing `jobArtifactDownloadAndDeserializeTime` metric.
3. ~~Misc. small updates for messages to improve clarity. Sanitize message for `logErrorApi` to remove potential AWS account Id~~

Tested metrics for all APIs are still emitted (except for small changes to metadata for `CreateUploadUrl` and `ExportResultArchive` as called out above). 

## Checklist
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [x] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt [hunk 1]
```diff
@@ -78,26 +78,11 @@ class ArtifactHandler(private val project: Project, private val clientAdaptor: G
...
-                    codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                    codeTransformApiErrorMessage = e.message.toString(),
-                    codeTransformJobId = job.id,
-                )
-                throw e // pass along error to callee
-            } finally {
-                apiEndTime = Instant.now()
-            }
+            LOG.info { "About to download the export result archive" }
+            val downloadResultsResponse = clientAdaptor.downloadExportResultArchive(job)
```
Comment (dhasani23) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt:82]: Do we not want to try-catch this anymore?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt [hunk 2]
```diff
@@ -106,33 +91,23 @@ class ArtifactHandler(private val project: Project, private val clientAdaptor: G
                     totalDownloadBytes += bytes.size
                 }
             }
-            CodetransformTelemetry.logApiLatency(
```
Comment (dhasani23) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt:109]: Why remove this?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt (hunk 1 above)
Comment (dhasani23) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/ArtifactHandler.kt:82]: I think I see. We're doing all of this in `GumbyClient.kt` now.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt [hunk 3]
```diff
@@ -37,64 +49,124 @@ class GumbyClient(private val project: Project) {
...
+                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
+                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
+                codeTransformUploadId = uploadId,
+                codeTransformJobId = jobId,
+                codeTransformRequestId = result?.responseMetadata()?.requestId(),
+            )
+        }
     }
 
     suspend fun downloadExportResultArchive(jobId: JobId): MutableList<ByteArray> {
```
Comment (dhasani23) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt:127]: Can we not call this API via the generic function just above?

Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt:127]: downloadExportResultArchive doesn't return a CodeWhispererRuntimeResponse, and has a lot more logic going on inside it, so I decided to keep it separate.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt [hunk 4]
```diff
@@ -37,64 +49,124 @@ class GumbyClient(private val project: Project) {
...
+                        .onBinaryMetadataEvent {
+                            checksum.set(it.contentChecksum())
+                        }.onBinaryPayloadEvent {
+                            val payloadBytes = it.bytes().asByteArray()
+                            byteBufferList.add(payloadBytes)
+                        }.onDefault {
+                            LOG.warn { "Received unknown payload stream: $it" }
+                        }
+                        .build()
+                )
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt:148]: you can still use the fluent builder for API request object
```suggestion
        val result = streamingBearerClient().exportResultArchive(
            {
                it.exportId(jobId.id)
                it.exportIntent(ExportIntent.TRANSFORMATION)
            },
            ExportResultArchiveResponseHandler.builder().subscriber(
...
...
```

 not really seeing an obvious nice way to write the handler visitor though

Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt:148]: it should match the `exportResultArchive(Consumer<ExportResultArchiveRequest.Builder> exportResultArchiveRequest, ExportResultArchiveResponseHandler asyncResponseHandler)` method signature and `it` is an instance of `ExportResultArchiveRequest.Builder`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerSession.kt [hunk 5]
```diff
@@ -525,6 +440,8 @@ class CodeModernizerSession(
         }
     }
 
+    private fun sanitizeMessage(error: String): String = error.replace(Regex("[0-9]{12}"), "{awsAccountId}")
```
[the remark this answers is not in the document]
Comment (tincheng) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerSession.kt]: Based on recent discussion in the gumby team, we may need additional filtering for messages that we want to include in metrics to remove various PII candidates. We may look into expanding this sanitization util further in future PRs.

**Comment by tincheng** [comment]
Based on latest discussion among the gumby team, we do not want to include full error message in metric. Latest commit remove the use of regex, and fixed some error logging to include original stack trace.

**Comment by tincheng** [comment]
This PR was from a while ago. We merged some other PRs which caused merge conflicts, and I have fixed the conflict now.

**Review by dhasani23 — APPROVED** [review]

**Review by rli — APPROVED** [review]

**Pull request merged**